DOCSP-43082: Cluster monitoring (#63)
mongodb/docs-c: source/monitoring/cluster-monitoring.txt

.. _c-cluster-monitoring:

==================
Cluster Monitoring
==================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, server, topology

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecols

Overview
--------

This guide shows you how to use the {+driver-short+} to monitor server
discovery and monitoring (SDAM) events in a MongoDB instance, replica
set, or sharded cluster. These events occur when there are any changes
in the state of the MongoDB instance or cluster that you are connected
to.

You might use information about SDAM events in your application to
understand cluster changes, assess cluster health, or perform capacity
planning.

.. _c-subscribe-sdam:

Subscribe to Events
-------------------

You can access details about SDAM events by subscribing to them
in your application. To subscribe to an event, create a value of
type ``mongoc_apm_callbacks_t`` and create an event callback function
for the event you want to subscribe to. Use the callback's corresponding
setter method to set the callback on the ``mongoc_apm_callbacks_t`` and 

> pass this type to the ``mongoc_client_set_apm_callbacks()`` function.

Reviewer: Recommended rewording:

> To subscribe to an event, define a callback function to handle each event type you want to subscribe to. Use a `mongoc_apm_callbacks_t` object to register the list of APM callbacks with a client object using `mongoc_client_set_apm_callbacks()`.

There is a lot of overlap with the actual code example below. Wordy details can be delegated to the code example.

Also recommend referring to these user-defined functions as "APM callbacks" or "APM callback functions" for consistency with C Driver API documentation.

This code monitors server opening events by performing the following
actions:

> - Defines a ``server_opening()`` event handler function, or callback

Reviewer: ```suggestion
- Defines a ``server_opening()`` APM callback function
```
For consistent terminology.

> - Creates a ``mongoc_apm_callbacks_t`` to store callbacks

Reviewer 2: ```suggestion
- Creates a ``mongoc_apm_callbacks_t`` instance to store callbacks
```

> - Creates a ``mongoc_apm_callbacks_t`` structure to store callbacks

Reviewer: ```suggestion
- Creates a ``mongoc_apm_callbacks_t`` object to store callbacks
```

> - Calls the ``mongoc_apm_set_server_opening_cb()`` function, which instructs
>   the ``mongoc_apm_callbacks_t`` type to store a pointer to the ``server_opening()``
>   callback

Reviewer: ```suggestion
- Calls the ``mongoc_apm_set_server_opening_cb()`` function, which stores a pointer to the provided APM callback function in the ``mongoc_apm_callbacks_t`` object
```

- Calls the ``mongoc_client_set_apm_callbacks()`` function, which registers

>   the callback with the client

Reviewer: ```suggestion
  the list of callbacks in the ``mongoc_apm_callbacks_t`` object with the client
```

Author: There's only one callback in this code example (server_opening()) rather than a list so I reworded slightly

.. literalinclude:: /includes/monitoring/sdam.c
   :language: c
   :copyable:
   :linenos:
   :emphasize-lines: 10, 30-32

When you perform a database operation, the driver establishes a new connection to
the server and your subscriber records the server opening event. The code outputs
messages that resemble the following:

.. code-block:: none
   :copyable: false

   Server opening on ac-rmuag0v-shard-00-00.gh0qg50.mongodb.net:27017
   Server opening on ac-rmuag0v-shard-00-01.gh0qg50.mongodb.net:27017

>    Server opening on ac-rmuag0v-shard-00-02.gh0qg50.mongodb.net:27017

Reviewer: These examples seem a bit noisy / unnecessarily specific ("ac-rmuag0v"? "gh0qg50"?). Can they be simplified?

Event Descriptions
------------------

You can subscribe to SDAM events by defining the corresponding

> event callback function. The following table provides the name of

Reviewer: ```suggestion
APM callback function. The following table provides the name of
```
Consistency.
